### PR TITLE
Report checkpoint processing to jcmd

### PR DESCRIPTION
**Pull request by Larry-N**
pass output stream from diagnosticCommand.cpp through java code into os_linux.cpp::VM_crac::doit()

---------
### Progress
- [x] Change must not contain extraneous whitespace

### Reviewers
 * [Anton Kozlov](https://openjdk.org/census#akozlov) (@AntonKozlov - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/crac pull/10/head:pull/10` \
`$ git checkout pull/10`

Update a local copy of the PR: \
`$ git checkout pull/10` \
`$ git pull https://git.openjdk.org/crac pull/10/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 10`

View PR using the GUI difftool: \
`$ git pr show -t 10`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/crac/pull/10.diff">https://git.openjdk.org/crac/pull/10.diff</a>

</details>

### CODE REVIEW

**Comment by Larry-N** [comment]
The example of output in jcmd:

`/home/source/git/crac/build/linux-x86_64-server-release/images/jdk/bin/jcmd target/example-jetty-1.0-SNAPSHOT.jar JDK.checkpoint
99007:
JDK.checkpoint command start processing
JVM: FD fd=0 type=character: details1="/dev/pts/0" OK: inherited from process env
JVM: FD fd=1 type=character: details1="/dev/pts/0" OK: inherited from process env
JVM: FD fd=2 type=character: details1="/dev/pts/0" OK: inherited from process env
JVM: FD fd=3 type=regular: details1="/home/source/git/crac/build/linux-x86_64-server-release/images/jdk/lib/modules" OK: inherited from process env
JVM: FD fd=4 type=regular: details1="/home/zulu-17381/example-jetty/target/example-jetty-1.0-SNAPSHOT.jar" OK: in classpath
JVM: FD fd=5 type=regular: details1="/home/zulu-17381/example-jetty/target/dependency/jetty-server-9.4.30.v20200611.jar" OK: assured persistent
JVM: FD fd=6 type=regular: details1="/home/zulu-17381/example-jetty/target/dependency/javax.servlet-api-3.1.0.jar" OK: assured persistent
JVM: FD fd=7 type=regular: details1="/home/zulu-17381/example-jetty/target/dependency/jetty-http-9.4.30.v20200611.jar" OK: assured persistent
JVM: FD fd=8 type=regular: details1="/home/zulu-17381/example-jetty/target/dependency/jetty-util-9.4.30.v20200611.jar" OK: assured persistent
JVM: FD fd=9 type=regular: details1="/home/zulu-17381/example-jetty/target/dependency/jetty-io-9.4.30.v20200611.jar" OK: assured persistent
JVM: FD fd=10 type=regular: details1="/home/zulu-17381/example-jetty/target/dependency/crac-999-20211028.191702-5.jar" OK: assured persistent
JVM: FD fd=23 type=socket: details1="socket:[7038828]" BAD: opened by application
 details2="socket:[7038828]" 
An exception durinng checkpoint operation: 
jdk.crac.impl.CheckpointOpenSocketException: socket:[7038828]
	at java.base/jdk.crac.Core.translateJVMExceptions(Core.java:90)
	at java.base/jdk.crac.Core.checkpointRestore1(Core.java:146)
	at java.base/jdk.crac.Core.checkpointRestore(Core.java:186)
	at java.base/jdk.crac.Core.checkpointRestoreInternal(Core.java:200)

JDK.checkpoint command processing finished
`

**Comment by Larry-N** [comment]
With the new changes, the opened socket (from jcmd) is checked "if the socket is from jcmd" - 
by parsing the /proc/<jcmd pid>/net/unix  file and getting inode number, comparing inode with file descriptor information in checkpoint processing ( function `VM_Crac::doit()` ) 
The opened socket is closed after writing all information regarding command processing just before calling the CRIU engine. 

An output example: 

```
oot@be23d3635404:/home/source/git/crac# /home/source/git/crac/build/linux-x86_64-server-release/images/jdk/bin/jcmd target/example-jetty-1.0-SNAPSHOT.jar JDK.checkpoint
72107:
JDK.checkpoint command start processing
JVM: FD fd=0 type=character: details1="/dev/pts/0" OK: inherited from process env
JVM: FD fd=1 type=character: details1="/dev/pts/0" OK: inherited from process env
JVM: FD fd=2 type=character: details1="/dev/pts/0" OK: inherited from process env
JVM: FD fd=3 type=regular: details1="/home/source/git/crac/build/linux-x86_64-server-release/images/jdk/lib/modules" OK: inherited from process env
JVM: FD fd=4 type=regular: details1="/home/zulu-17381/example-jetty/target/example-jetty-1.0-SNAPSHOT.jar" OK: in classpath
JVM: FD fd=5 type=regular: details1="/home/zulu-17381/example-jetty/target/dependency/jetty-server-9.4.30.v20200611.jar" OK: assured persistent
JVM: FD fd=6 type=regular: details1="/home/zulu-17381/example-jetty/target/dependency/javax.servlet-api-3.1.0.jar" OK: assured persistent
JVM: FD fd=7 type=regular: details1="/home/zulu-17381/example-jetty/target/dependency/jetty-http-9.4.30.v20200611.jar" OK: assured persistent
JVM: FD fd=8 type=regular: details1="/home/zulu-17381/example-jetty/target/dependency/jetty-util-9.4.30.v20200611.jar" OK: assured persistent
JVM: FD fd=9 type=regular: details1="/home/zulu-17381/example-jetty/target/dependency/jetty-io-9.4.30.v20200611.jar" OK: assured persistent
JVM: FD fd=10 type=regular: details1="/home/zulu-17381/example-jetty/target/dependency/crac-999-20211028.191702-5.jar" OK: assured persistent
JVM: FD fd=23 type=socket: details1="socket:[735739]" issock, details2="socket:[735739]" OK: jcmd socket
CR: CR Checkpoint ...

```

File: src/hotspot/os/linux/os_linux.cpp [hunk 1]
```diff
@@ -6190,67 +6216,63 @@ void VM_Crac::doit() {
...
-    ok = false;
-
     if (S_ISSOCK(st->st_mode)) {
-      details = sock_details(details, detailsbuf, sizeof(detailsbuf));
-      if (CRPrintResourcesOnCheckpoint) {
-        tty->print(" details2=\"%s\" ", details);
+      if (is_socket_from_jcmd(i)){
+        print_resources(ostream, "OK: jcmd socket");
+        ok = true;
+        continue;
```
Comment (AntonKozlov) [inline src/hotspot/os/linux/os_linux.cpp:6238]: You should not reset `ok` state here -- continue below should be enough. Otherwise you may lose info about e.g. previously found socket.

File: src/hotspot/os/linux/os_linux.cpp [hunk 2]
```diff
@@ -6190,67 +6216,63 @@ void VM_Crac::doit() {
...
 
-  if (!ok && CRHeapDumpOnCheckpointException) {
-    HeapDumper::dump_heap();
-  }
-
-  if (!ok && CRDoThrowCheckpointException) {
+  if (!ok) {
+    if (CRHeapDumpOnCheckpointException){
+      HeapDumper::dump_heap();
+    }
```
Comment (AntonKozlov) [inline src/hotspot/os/linux/os_linux.cpp:6255]: What is the reason for this change? CRDoThrowCheckpointException is useful sometimes. Please revert back.

File: src/hotspot/os/linux/os_linux.cpp [hunk 3]
```diff
@@ -6361,6 +6383,8 @@ Handle os::Linux::checkpoint(bool dry_run, TRAPS) {
 
   VM_Crac cr(dry_run);
   {
+    cr.jcmd_operation = (LinuxAttachOperation *) op;
+    cr.ostream = (outputStream *) stream;
```
Comment (AntonKozlov) [inline src/hotspot/os/linux/os_linux.cpp:6387]: Please initialize these in the constructor.

File: src/java.base/share/classes/jdk/crac/Core.java [hunk 4]
```diff
@@ -247,23 +251,34 @@ public static void checkpointRestore() throws
...
+            // jcmd + somebody through an api ...
+            if (!checkpointInProgress) {
+                checkpointInProgress = true;
+                try {
+                    checkpointRestore1(outputStream_p, jcmd_p);
+                } finally {
+                    checkpointInProgress = false;
+                }
             }
+        }
```
Comment (AntonKozlov) [inline src/java.base/share/classes/jdk/crac/Core.java:268]: Let's move the whole recursive check from public checkpointRestore() here, and make that only to call the private method.

**Review by AntonKozlov — CHANGES_REQUESTED** [review]
Few notes from my side. I haven't run this yet, will try in a couple of days. Thanks!

File: src/hotspot/os/linux/attachListener_linux.cpp [hunk 5]
```diff
@@ -405,11 +353,23 @@ int LinuxAttachListener::write_fully(int s, char* buf, int len) {
 // if there are operations that involves a very big reply then it the
 // socket could be made non-blocking and a timeout could be used.
 
-void LinuxAttachOperation::complete(jint result, bufferedStream* st) {
-  JavaThread* thread = JavaThread::current();
-  ThreadBlockInVM tbivm(thread);
+void LinuxAttachOperation::effectively_complete_raw(jint result, bufferedStream* st) {
+
+  if (_effectively_completed) {
+    return;
```
Comment (AntonKozlov) [inline src/hotspot/os/linux/attachListener_linux.cpp:359]: An assert here would catch lost output that won't be reported anywhere.
```suggestion
    assert(st->size() == 0, "no lost output");
    return;
```

File: src/hotspot/share/services/diagnosticCommand.cpp [hunk 6]
```diff
@@ -1040,9 +1041,20 @@ void DebugOnCmdStartDCmd::execute(DCmdSource source, TRAPS) {
 void CheckpointDCmd::execute(DCmdSource source, TRAPS) {
   Klass* k = SystemDictionary::resolve_or_fail(vmSymbols::jdk_crac_Core(),
                                                  true, CHECK);
-  JavaValue result(T_VOID);
+  JavaValue result(T_OBJECT);
+  JavaCallArguments args;
+  args.push_long((jlong )output());
+  args.push_long((jlong )LinuxAttachListener::get_jcmdOperation());
```
Comment (AntonKozlov) [inline src/hotspot/share/services/diagnosticCommand.cpp:1047]: You apparently don't need to pass the attach operation through java code to JVM_Checkpoint and eventually os::Linux::checkpoint -- you can get the operation right there with the same call.

**Review by AntonKozlov — CHANGES_REQUESTED** [review]

File: src/hotspot/os/linux/os_linux.cpp [hunk 7]
```diff
@@ -6243,14 +6263,19 @@ void VM_Crac::doit() {
...
-  int ret = checkpoint_restore(&shmid);
-  if (ret == JVM_CHECKPOINT_ERROR) {
-    PerfMemoryLinux::restore();
-    return;
+  if (CRAllowToSkipCheckpoint) {
+    trace_cr(ostream, "Skip Checkpoint");
+  } else {
+    trace_cr(ostream, "Checkpoint ...");
+    // If execution comes here, assumme that further all be ok.
+    report_ok_to_jcmd();
```
[the remark this answers is not in the document]
Comment (Larry-N) [reply src/hotspot/os/linux/os_linux.cpp:6272]: Further it goes to `tty` now: 

```
static int checkpoint_restore(int *shmid) {
...
  if (CRTraceStartupTime) {
    tty->print_cr("STARTUPTIME " JLONG_FORMAT " restore-native", os::javaTimeNanos());
  }

  if (info.si_code != SI_QUEUE || info.si_int < 0) {
    tty->print("JVM: invalid info for restore provided: %s", info.si_code == SI_QUEUE ? "queued" : "not queued");
    if (info.si_code == SI_QUEUE) {
      tty->print(" code %d", info.si_int);
...
```

File: src/hotspot/os/linux/os_linux.cpp [hunk 8]
```diff
@@ -5706,12 +5716,23 @@ bool os::supports_map_sync() {
...
+    va_list ap;
+    va_start(ap, msg);
+    ou->print("CR: ");
+    ou->vprint_cr(msg, ap);
+    va_end(ap);
+  }
+}
+
+static void print_resources(outputStream * ostream, const char* msg, ... ) {
+  outputStream * ou = (ostream == NULL) ? tty : ostream;
```
[the remark this answers is not in the document]
Comment (Larry-N) [reply src/hotspot/os/linux/os_linux.cpp:5731]: Moved the output stream selection into VM_Crac constructor, but propose still use these functions for hiding check on `CRPrintResourcesOnCheckpoint`,  `CRTrace`.

File: src/hotspot/os/linux/attachListener_linux.cpp (hunk 5 above)
Comment (Larry-N) [reply src/hotspot/os/linux/attachListener_linux.cpp:359]: A straightforward implementation of this assertion leads to triggering it at restore. 
Seems like `LinuxAttachListener::write_fully()` doesnt reset the buffer_pos in `bufferedStream`, so the `bufferedStream->size() `is non-zero.  

Probably, it worth to add `bufferedStream->reset()` to `LinuxAttachOperation::write_operation_result()`

Comment (AntonKozlov) [reply src/hotspot/os/linux/attachListener_linux.cpp:359]: I believe the data there is the text I observe with the CRAllowToSkipCheckpoint. So the root casue should be fixed (the text that won't be shown).

Comment (Larry-N) [reply src/hotspot/os/linux/attachListener_linux.cpp:362]: Looks like the issue with this assertion was fixed with the commit 29365804b4be9d0cf41a4feaf5ac46fe823d4fba

**Review by AntonKozlov — APPROVED** [review]

**Comment by AntonKozlov** [comment]
I've played with the most recent version of this change and it works as expected :+1: 

Interesting that VM messages go to the jcmd stream, and probably we may want Resource to be able to log to that stream as well. It may be a follow-up for this change.

BTW, have you found the reason the submit test has failed?

**Comment by Larry-N** [comment]
No, I didn't find the reason yet
Referred to this link ( that passed ok )it looks like tests hs/tier1_serviceability is unstable for some reason
https://github.com/Larry-N/crac/actions/runs/3387045467

**Comment by AntonKozlov** [comment]
OK, I see it has passed.

**Review by AntonKozlov — APPROVED** [review]
Still looks fine.
